### PR TITLE
SYCL: support 0-dim acc in handler::copy(accessor, accessor)

### PR DESCRIPTION
**Pull request by v-klochkov**
Signed-off-by: Vyacheslav N Klochkov <vyacheslav.n.klochkov@intel.com>

### CODE REVIEW

File: sycl/include/CL/sycl/handler.hpp [hunk 1]
```diff
@@ -366,6 +367,109 @@ class __SYCL_EXPORT handler {
...
+      Dst[DstIndex] = Src[SrcIndex];
+    });
+    return true;
+  }
+
+  template <typename T, int Dim, access::mode Mode, access::target Target,
+            access::placeholder IsPH>
+  detail::enable_if_t<Dim == 0 && Mode == access::mode::atomic, T>
+  readFromFirstAccElement(accessor<T, Dim, Mode, Target, IsPH> Src) const {
+    auto AtomicSrc = (atomic<T, access::address_space::global_space>)Src;
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/handler.hpp:405]: Why do we need to explicitly convert here? The spec says that operator[] of accessor with access mode == atomic returns object of atomic class.

File: sycl/include/CL/sycl/handler.hpp [hunk 2]
```diff
@@ -366,6 +367,109 @@ class __SYCL_EXPORT handler {
...
+    return AtomicSrc.read();
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/handler.hpp:406]: Can't find read method of atomic class. Am I missing something?

File: sycl/include/CL/sycl/handler.hpp (hunk 1 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:405]: Operator [] is available only when Dim > 0:
97 /* Available only when: accessMode == access::mode::atomic && dimensions >
98 0 */
99 atomic<dataT, access::address_space::global_space> operator[](
100 id<dimensions> index) const;

The cast is needed when Dim==0

File: sycl/include/CL/sycl/handler.hpp (hunk 2 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:406]: You're right, it is not read(), it is load() method.
I need to update the LIT test to check that properly.

**Comment by s-kanaev** [comment]
@v-klochkov  Please apply clang-format to the patch. You can find usefule artifacts in deails of clang-format-check job details.

**Comment by v-klochkov** [comment]
@s-kanaev I gave some honor to clang-format and added few additional/recommended changes.
I though want to ignore few places 
a) 2 places in handler.hpp: the code without clang-format is much more readable and 
b) 2 places in LIT test: I just follow the more compact style of all other similar funcs in LIT test.

**Comment by v-klochkov** [comment]
@romanovvlad Thank you for the quick initial review.
I updated the changes in handler.hpp and also added 3 more cases to LIT test to check atomics.

File: sycl/include/CL/sycl/handler.hpp [hunk 3]
```diff
@@ -366,6 +367,107 @@ class __SYCL_EXPORT handler {
...
+  /// to another accessor. Returns true if the copy is handled here.
+  ///
+  /// \param Src is a source SYCL accessor.
+  /// \param Dst is a destination SYCL accessor.
+  // TODO: support atomic accessor in Src or/and Dst.
+  template<
+      typename TSrc, int DimSrc, access::mode ModeSrc, access::target TargetSrc,
+      typename TDst, int DimDst, access::mode ModeDst, access::target TargetDst,
+      access::placeholder IsPHSrc, access::placeholder IsPHDst>
+  detail::enable_if_t<(DimSrc > 0) && (DimDst > 0), bool>
```
Comment (s-kanaev) [inline sycl/include/CL/sycl/handler.hpp:380]: Should we check here if `DimSrc == DimDst`?

File: sycl/include/CL/sycl/handler.hpp [hunk 4]
```diff
@@ -366,6 +367,107 @@ class __SYCL_EXPORT handler {
...
+      Dst[DstIndex] = Src[SrcIndex];
+    });
+    return true;
+  }
+
+  template <typename T, int Dim, access::mode Mode, access::target Target,
+            access::placeholder IsPH>
+  detail::enable_if_t<Dim == 0 && Mode == access::mode::atomic, T>
+  readFromFirstAccElement(accessor<T, Dim, Mode, Target, IsPH> Src) const {
+    auto AtomicSrc = (atomic<T, access::address_space::global_space>)Src;
```
Comment (s-kanaev) [inline sycl/include/CL/sycl/handler.hpp:404]: Could this conversion be converted to this?

```suggestion
    atomic<T, access::address_space::global_space> AtomicSrc = Src;
```

File: sycl/include/CL/sycl/handler.hpp [hunk 5]
```diff
@@ -366,6 +367,107 @@ class __SYCL_EXPORT handler {
...
+  detail::enable_if_t<Mode != access::mode::atomic, T>
+  readFromFirstAccElement(accessor<T, Dim, Mode, Target, IsPH> Src) const {
+    return *(Src.get_pointer());
+  }
+
+  template <typename T, int Dim, access::mode Mode, access::target Target,
+            access::placeholder IsPH>
+  detail::enable_if_t<Dim == 0 && Mode == access::mode::atomic, void>
+  writeToFirstAccElement(accessor<T, Dim, Mode, Target, IsPH> Dst, T V) const {
+    auto AtomicDst = (atomic<T, access::address_space::global_space>)Dst;
```
Comment (s-kanaev) [inline sycl/include/CL/sycl/handler.hpp:427]: Same thing here

**Comment by s-kanaev** [comment]
> @s-kanaev I gave some honor to clang-format and added few additional/recommended changes.
I though want to ignore few places
a) 2 places in handler.hpp: the code without clang-format is much more readable and
b) 2 places in LIT test: I just follow the more compact style of all other similar funcs in LIT test.

I believe you have to massage the code to make clang-format happy.

File: sycl/include/CL/sycl/handler.hpp (hunk 3 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:380]: I don't know if it is correct to copy from 1D accessor to 3D. It is probably correct. So, that check is not needed.
In my patch I do not change any constraints. 
My patch only separates the existing code to this routine to make it possible to copy to/from 0-dim accessor.

File: sycl/include/CL/sycl/handler.hpp (hunk 4 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp]: Ok, I changed it that way.

File: sycl/include/CL/sycl/handler.hpp (hunk 5 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp]: Fixed.

**Comment by v-klochkov** [comment]
@s-kanaev 

> I believe you have to massage the code to make clang-format happy.

I already applied clang-format per your request, and ignored format's recommendation in only 4 places in sake of better code readability. Did you read my reasoning above?
Sure, clang-format is good in 95%, but let's not turn it into dogma, please.

**Comment by s-kanaev** [comment]
> I already applied clang-format per your request, and ignored format's recommendation in only 4 places in sake of better code readability. Did you read my reasoning above?
Sure, clang-format is good in 95%, but let's not turn it into dogma, please.

I understand readability as a pretty good reason. The justification for my argument is that I'm not really sure that "Merge" button will be available with clang-format job failed.

File: sycl/include/CL/sycl/handler.hpp [hunk 6]
```diff
@@ -366,6 +367,107 @@ class __SYCL_EXPORT handler {
...
+  detail::enable_if_t<Mode != access::mode::atomic, T>
+  readFromFirstAccElement(accessor<T, Dim, Mode, Target, IsPH> Src) const {
+    return *(Src.get_pointer());
+  }
+
+  template <typename T, int Dim, access::mode Mode, access::target Target,
+            access::placeholder IsPH>
+  detail::enable_if_t<Dim == 0 && Mode == access::mode::atomic, void>
+  writeToFirstAccElement(accessor<T, Dim, Mode, Target, IsPH> Dst, T V) const {
+    atomic<T, access::address_space::global_space> AtomicDst = Dst;
```
Comment (s-kanaev) [inline sycl/include/CL/sycl/handler.hpp:427]: Shouldn't it be a reference to atomic?

**Comment by romanovvlad** [comment]
> 
> 
> > I already applied clang-format per your request, and ignored format's recommendation in only 4 places in sake of better code readability. Did you read my reasoning above?
> > Sure, clang-format is good in 95%, but let's not turn it into dogma, please.
> 
> I understand readability as a pretty good reason. The justification for my argument is that I'm not really sure that "Merge" button will be available with clang-format job failed.

clang-format-check is not required so we should be able to merge even if it fails

**Comment by bader** [comment]
> > I already applied clang-format per your request, and ignored format's recommendation in only 4 places in sake of better code readability. Did you read my reasoning above?
> > Sure, clang-format is good in 95%, but let's not turn it into dogma, please.
> 
> I understand readability as a pretty good reason. The justification for my argument is that I'm not really sure that "Merge" button will be available with clang-format job failed.

The reason to make this change is to make consistent formatting across the project, not to make "merge button available".

@v-klochkov, if you think that our project should not accept changes proposed by clang-format, please suggest clang-format configuration file changes. It's much better approach than just break the guidelines followed by the community.

**Comment by s-kanaev** [comment]
> The reason to make this change is to make consistent formatting across the project, not to make "merge button available".

I understand it pretty much. I have only translated mechanics or sequence of actions in plain English.

File: sycl/include/CL/sycl/handler.hpp (hunk 6 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:427]: The accessor operator[] and operator() return a value, not a reference.

93 /* Available only when: accessMode == access::mode::atomic && dimensions ==
94 0 */
95 operator atomic<dataT, access::address_space::global_space> () const;
96
97 /* Available only when: accessMode == access::mode::atomic && dimensions >
98 0 */
99 atomic<dataT, access::address_space::global_space> operator[](
100 id<dimensions> index) const;

Please clarify if I understood your comment incorrectly.
Did you think about adding & sign right before AtomicDst?
atomic<T, access::address_space::global_space> &AtomicDst = Dst;
If so it would be incorrect code, I think.

**Comment by v-klochkov** [comment]
> > I understand readability as a pretty good reason. The justification for my argument is that I'm not really sure that "Merge" button will be available with clang-format job failed.
> 
> The reason to make this change is to make consistent formatting across the project, not to make "merge button available".
> 
> @v-klochkov, if you think that our project should not accept changes proposed by clang-format, please suggest clang-format configuration file changes. It's much better approach than just break the guidelines followed by the community.

Teaching clang-format to recognize the existing laconic/1-line style/format already used in file,
or teaching it to recognize pattern in arg names does not seem easy doable and/or deserving time.

'clang-format' check did not have "Required" status, thus I thought there was some flexibility here.
The clang-format issue is fixed now.
